app/controllers/v1/submission_entries_controller.rb

```diff
@@ -5,6 +5,8 @@ def create
     entry = initialize_submission_entry
     entry.attributes = submission_entry_params
 
+    return head :conflict if SubmissionEntry.exists?(submission_id: entry.submission_id, source: entry.source)
```

Comment:
This might need an index; try `SubmissionEntry.where(submission_id: entry.submission_id, source: entry.source)` and it's doing a `Bitmap Heap Scan` on `submission_entries`. That means its found all the entries for the submission (on my old DB cut there are about 300 of them) and it'll have to compare the whole JSON column in a bitwise comparison sense. I *think* that'll be expensive but I'm having trouble translating this to a raw PSQL query because I don't know enough of the JSONB syntax to `WHERE` a whole field in the same sense so I can't get to `EXPLAIN ANALYZE`.

If you fancy pairing with a screen share give me a shout – we might both be able to learn something about Postgres and GIN indexes :)

Reply:
Oh hang on we've got one: `"index_submission_entries_on_source" gin (source)`. Tekin added one in a migration that's post- my DB dump. I wonder why it's not being used?

Reply:
If I run the query on staging it seems pretty fast - I'm after all matching on the entirety of the source column, and there's a GIN index on it. 

We can probably make this more performant with additional indexes, but this is only here as a quick fix, so not sure if the additional work is worth it at the moment...